planner/core/expression_rewriter.go

```diff
@@ -762,21 +769,36 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
 		er.err = expression.ErrOperandColumns.GenWithStackByArgs(lLen)
 		return v, true
 	}
-	var rexpr expression.Expression
-	if np.Schema().Len() == 1 {
-		rexpr = np.Schema().Columns[0]
-		rCol := rexpr.(*expression.Column)
-		// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
-		// normal column equal condition, so we specially mark the inner operand here.
-		if v.Not || asScalar {
+	// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
+	// normal column equal condition, so we specially mark the inner operand here.
+	if v.Not || asScalar {
+		for i, rCol := range np.Schema().Columns {
+			larg := lexpr
+			lSf, isSf := lexpr.(*expression.ScalarFunction)
+			if isSf {
+				larg = lSf.GetArgs()[i]
+			}
 			rCol.InOperand = true
 			// If both input columns of `in` expression are not null, we can treat the expression
 			// as normal column equal condition instead.
-			lCol, ok := lexpr.(*expression.Column)
+			lCol, ok := larg.(*expression.Column)
 			if ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 				rCol.InOperand = false
 			}
+			lCon, ok := larg.(*expression.Constant)
+			if ok && !lCon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+				rCol.InOperand = false
+			}
+			lClc, ok := larg.(*expression.CorrelatedColumn)
+			if ok && !lClc.Data.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+				rCol.InOperand = false
+			}
```

Comment:
Extract this part of logic as a single function, and reuse it in `buildSemiApplyFromEqualSubq` as well?

Reply:
fixed